spec/decorators/address_form_decorator_spec.rb

```diff
@@ -0,0 +1,118 @@
+require 'rails_helper'
+
+RSpec.describe AddressFormDecorator do
+  subject { described_class.new(form_object) }
+
+  let(:form_object) { double('FormObject', record: address_record) }
+
+  let(:address_record) { instance_double(Address, type: address_type, person: person_record) }
+  let(:person_record)  { instance_double(Person, type: person_type) }
+
+  let(:address_type) { nil }
+  let(:person_type)  { nil }
+
+  # NOTE: we only really have for now `Applicant`, but as these decorator is built
+  # to cope with multiple people types, we are testing it also for `Partner`
+
+  describe '#page_title' do
+    context 'for an applicant' do
+      let(:person_type) { Applicant.to_s }
+
+      context 'for a home address' do
+        let(:address_type) { HomeAddress.to_s }
+        it { expect(subject.page_title).to eq('.page_title.applicant.home_address') }
+      end
+
+      context 'for a correspondence address' do
+        let(:address_type) { CorrespondenceAddress.to_s }
+        it { expect(subject.page_title).to eq('.page_title.applicant.correspondence_address') }
+      end
+    end
+
+    context 'for a partner' do
+      let(:person_type) { Partner.to_s }
+
+      context 'for a home address' do
+        let(:address_type) { HomeAddress.to_s }
+        it { expect(subject.page_title).to eq('.page_title.partner.home_address') }
+      end
+
+      context 'for a correspondence address' do
+        let(:address_type) { CorrespondenceAddress.to_s }
+        it { expect(subject.page_title).to eq('.page_title.partner.correspondence_address') }
+      end
+    end
+  end
+
+  describe '#heading' do
+    context 'for an applicant' do
+      let(:person_type) { Applicant.to_s }
+
+      context 'for a home address' do
+        let(:address_type) { HomeAddress.to_s }
+        it { expect(subject.heading).to eq('.heading.applicant.home_address') }
+      end
+
+      context 'for a correspondence address' do
+        let(:address_type) { CorrespondenceAddress.to_s }
+        it { expect(subject.heading).to eq('.heading.applicant.correspondence_address') }
+      end
+    end
+
+    context 'for a partner' do
+      let(:person_type) { Partner.to_s }
+
+      context 'for a home address' do
+        let(:address_type) { HomeAddress.to_s }
+        it { expect(subject.heading).to eq('.heading.partner.home_address') }
+      end
+
+      context 'for a correspondence address' do
```

Comment:
Likewise here. Obviously no issue to leave in, but could be misleading to anyone reading the specs in the future as to what the functionality should be

Reply:
Yep, same as comment above.

We really want to test all scenarios this decorator can cope with. The code has been made generic on purpose to cope with any combination, and the decorator doesn't have any business logic so if then we don't have correspondence address for partner that is all fine as that's a business decision (we just don't show that journey to the user) but internally the code has to work for any type, as we could even have addresses for the codefendants if we wanted too.

Reply:
Ok, sounds good